### ISSUE TITLE
On the dev feature if_any/if_all (this is not a bug)

### ISSUE THREAD

**Issue by iago-pssjd** [tidyverse/dplyr#5709]
Hi!
I just installed the dev version of `dplyr` to test the new feature produced after #5660.

I run the examples using the next data:
```r
library(dplyr)
library(tibble)
df <- data.frame(
    "ID" = c("asdf","vfdkun", "seifu", "seijd", "qweri"),
    "elephant_zoo" = c(1,1,1,2,0), #Should not be useful there
    "rhino_zoo" = c(1,2,3,1,0), #Should not be useful there
    "hippo_zoo" = c(1,1,0,0,0),
    "elephant_wild_A" = c(0,0,1,1,3),
    "rhino_wild_A" = c(0,0,4,3,1),
    "elephant_wild_B" = c(0,0,0,0,0),
    "rhino_wild_C" = c(0,0,1,5,7),
    "hippo_wild_B" = c(0,0,0,0,0)) %>% 
  column_to_rownames(var = "ID")
df
       elephant_zoo rhino_zoo hippo_zoo elephant_wild_A rhino_wild_A elephant_wild_B rhino_wild_C
asdf              1         1         1               0            0               0            0
vfdkun            1         2         1               0            0               0            0
seifu             1         3         0               1            4               0            1
seijd             2         1         0               1            3               0            5
qweri             0         0         0               3            1               0            7
       hippo_wild_B
asdf              0
vfdkun            0
seifu             0
seijd             0
qweri             0
```
It works as expected, since it works inside `filter`:
```r
df %>% filter(if_any(contains("zoo"), ~`>`(.,0)))
       elephant_zoo rhino_zoo hippo_zoo elephant_wild_A rhino_wild_A elephant_wild_B rhino_wild_C
asdf              1         1         1               0            0               0            0
vfdkun            1         2         1               0            0               0            0
seifu             1         3         0               1            4               0            1
seijd             2         1         0               1            3               0            5
       hippo_wild_B
asdf              0
vfdkun            0
seifu             0
seijd             0
```
But I hoped that it worked beyond that, which is not the case:
```r
df %>%
    mutate(ZOO = 
               case_when(if_any(contains("zoo"), ~`>`(.,0)) ~ "zoo", 
                         TRUE ~ "F"))
Error: Problem with `mutate()` input `ZOO`.
x LHS of case 1 (`if_any(contains("zoo"), ~. > 0)`) must be a logical vector, not a `tbl_df/tbl/data.frame` object.
ℹ Input `ZOO` is `case_when(if_any(contains("zoo"), ~. > 0) ~ "zoo", TRUE ~ "F")`.
Run `rlang::last_error()` to see where the error occurred.
```
So not a single logical vector outside `filter`... Let's try
```r
df %>%
    mutate(ZOO = 
               case_when(pull(if_any(contains("zoo"), ~`>`(.,0))) ~ "zoo", 
                         TRUE ~ "F"))
       elephant_zoo rhino_zoo hippo_zoo elephant_wild_A rhino_wild_A elephant_wild_B rhino_wild_C hippo_wild_B ZOO
asdf              1         1         1               0            0               0            0            0 zoo
vfdkun            1         2         1               0            0               0            0            0 zoo
seifu             1         3         0               1            4               0            1            0   F
seijd             2         1         0               1            3               0            5            0   F
qweri             0         0         0               3            1               0            7            0   F
```
The result is incorrect. Indeed:
```r
df %>%
    mutate(ZOO = pull(if_any(contains("zoo"), ~`>`(.,0))))
       elephant_zoo rhino_zoo hippo_zoo elephant_wild_A rhino_wild_A elephant_wild_B rhino_wild_C hippo_wild_B   ZOO
asdf              1         1         1               0            0               0            0            0  TRUE
vfdkun            1         2         1               0            0               0            0            0  TRUE
seifu             1         3         0               1            4               0            1            0 FALSE
seijd             2         1         0               1            3               0            5            0 FALSE
qweri             0         0         0               3            1               0            7            0 FALSE
```
Let's see:
```r
df %>%
    mutate(ZOO = if_any(contains("zoo"), ~`>`(.,0)))
       elephant_zoo rhino_zoo hippo_zoo elephant_wild_A rhino_wild_A elephant_wild_B rhino_wild_C hippo_wild_B
asdf              1         1         1               0            0               0            0            0
vfdkun            1         2         1               0            0               0            0            0
seifu             1         3         0               1            4               0            1            0
seijd             2         1         0               1            3               0            5            0
qweri             0         0         0               3            1               0            7            0
                                                                                                                                                                           ZOO
asdf                                                                                                                                   \033[38;5;246m# A tibble: 5 x 3\033[39m
vfdkun                                                                                                                                        elephant_zoo rhino_zoo hippo_zoo
seifu  \033[38;5;250m*\033[39m \033[3m\033[38;5;246m<lgl>\033[39m\033[23m        \033[3m\033[38;5;246m<lgl>\033[39m\033[23m     \033[3m\033[38;5;246m<lgl>\033[39m\033[23m    
seijd                                                                                                                 \033[38;5;250m1\033[39m TRUE         TRUE      TRUE     
qweri                                                                                                                 \033[38;5;250m2\033[39m TRUE         TRUE      TRUE     
Warning message:
In format.data.frame(if (omit) x[seq_len(n0), , drop = FALSE] else x,  :
  corrupt data frame: columns will be truncated or padded with NAs
```
Ough!!!

It would be nice to have these functions beyond `filter`. Meanwhile, we wil go on using `any_cols` as in #4770.

Thank you!

**Comment by romainfrancois** [comment]
Interesting. The way we've implemented `if_any()` so far is to add an extra bit of information to the data frame made by `across()` to signal to `filter()` that it should be "reduced" with `or()` : 

``` r
dplyr::if_any
#> function(.cols, .fns = NULL, ..., .names = NULL) {
#>   structure(
#>     across({{.cols}}, .fns = .fns, ..., .names = .names),
#>     filter_combine = "or"
#>   )
#> }
#> <bytecode: 0x7f9766c726c0>
#> <environment: namespace:dplyr>
```

<sup>Created on 2021-01-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

It might be more strategic if `if_any()` would do the reducing itself: 

``` r
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(tibble)

df <- data.frame(
  "ID" = c("asdf","vfdkun", "seifu", "seijd", "qweri"),
  "elephant_zoo" = c(1,1,1,2,0), #Should not be useful there
  "rhino_zoo" = c(1,2,3,1,0), #Should not be useful there
  "hippo_zoo" = c(1,1,0,0,0),
  "elephant_wild_A" = c(0,0,1,1,3),
  "rhino_wild_A" = c(0,0,4,3,1),
  "elephant_wild_B" = c(0,0,0,0,0),
  "rhino_wild_C" = c(0,0,1,5,7),
  "hippo_wild_B" = c(0,0,0,0,0)) %>% 
  column_to_rownames(var = "ID")

df %>% filter(if_any(contains("zoo"), ~`>`(.,0)))
#>        elephant_zoo rhino_zoo hippo_zoo elephant_wild_A rhino_wild_A
#> asdf              1         1         1               0            0
#> vfdkun            1         2         1               0            0
#> seifu             1         3         0               1            4
#> seijd             2         1         0               1            3
#>        elephant_wild_B rhino_wild_C hippo_wild_B
#> asdf                 0            0            0
#> vfdkun               0            0            0
#> seifu                0            1            0
#> seijd                0            5            0
df %>%
  mutate(ZOO = {
    xx <- if_any(contains("zoo"), ~`>`(.,0))
    str(xx)
    42
  }) -> forget
#> tibble [5 × 3] (S3: tbl_df/tbl/data.frame)
#>  $ elephant_zoo: logi [1:5] TRUE TRUE TRUE TRUE FALSE
#>  $ rhino_zoo   : logi [1:5] TRUE TRUE TRUE TRUE FALSE
#>  $ hippo_zoo   : logi [1:5] TRUE TRUE FALSE FALSE FALSE
#>  - attr(*, "filter_combine")= chr "or"

df %>% filter(if_any(contains("zoo"), ~`>`(.,0)))
#>        elephant_zoo rhino_zoo hippo_zoo elephant_wild_A rhino_wild_A
#> asdf              1         1         1               0            0
#> vfdkun            1         2         1               0            0
#> seifu             1         3         0               1            4
#> seijd             2         1         0               1            3
#>        elephant_wild_B rhino_wild_C hippo_wild_B
#> asdf                 0            0            0
#> vfdkun               0            0            0
#> seifu                0            1            0
#> seijd                0            5            0
df %>%
  mutate(ZOO = {
    xx <- purrr::reduce(if_any(contains("zoo"), ~`>`(.,0)), `|`)
    str(xx)
    case_when(xx ~ "zoo", TRUE ~ "F")
  })
#>  logi [1:5] TRUE TRUE TRUE TRUE FALSE
#>        elephant_zoo rhino_zoo hippo_zoo elephant_wild_A rhino_wild_A
#> asdf              1         1         1               0            0
#> vfdkun            1         2         1               0            0
#> seifu             1         3         0               1            4
#> seijd             2         1         0               1            3
#> qweri             0         0         0               3            1
#>        elephant_wild_B rhino_wild_C hippo_wild_B ZOO
#> asdf                 0            0            0 zoo
#> vfdkun               0            0            0 zoo
#> seifu                0            1            0 zoo
#> seijd                0            5            0 zoo
#> qweri                0            7            0   F
```

<sup>Created on 2021-01-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
